skillner/utils.py

```diff
@@ -0,0 +1,477 @@
+"""
```

Comment:
due to the PR comment, i haven't looked at this in a lot of detail, but broadly looks great anyway!